github/transport.go

```diff
@@ -29,3 +38,119 @@ func (ett *etagTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+
+	m sync.Mutex
+}
+
+func (rlt *rateLimitTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Make requests for a single user or client ID serially
+	// This is also necessary for safely saving
+	// and restoring bodies between retries below
+	rlt.lock(req)
+
+	// If you're making a large number of POST, PATCH, PUT, or DELETE requests
+	// for a single user or client ID, wait at least one second between each request.
+	if rlt.delayNextRequest {
+		log.Printf("[DEBUG] Sleeping %s between write operations", writeDelay)
+		time.Sleep(writeDelay)
+	}
+	if rlt.isWriteMethod(req.Method) {
+		rlt.delayNextRequest = true
+	} else {
+		rlt.delayNextRequest = false
+	}
+
+	resp, err := rlt.transport.RoundTrip(req)
+	if err != nil {
+		rlt.unlock(req)
+		return resp, err
+	}
+
+	// Make response body accessible for retries & debugging
+	// (work around bug in GitHub SDK)
+	// See https://github.com/google/go-github/pull/986
+	r1, r2, err := rlt.drainBody(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	resp.Body = r1
+	ghErr := github.CheckResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+	resp.Body = r2
+
+	// When you have been limited, use the Retry-After response header to slow down.
+	if arlErr, ok := ghErr.(*github.AbuseRateLimitError); ok {
+		rlt.delayNextRequest = false
+		retryAfter := arlErr.GetRetryAfter()
+		log.Printf("[DEBUG] Abuse detection mechanism triggered, sleeping for %s before retrying",
+			retryAfter)
+		time.Sleep(retryAfter)
+		rlt.unlock(req)
+		return rlt.RoundTrip(req)
+	}
+
+	if rlErr, ok := ghErr.(*github.RateLimitError); ok {
+		rlt.delayNextRequest = false
+		retryAfter := rlErr.Rate.Reset.Sub(time.Now())
+		log.Printf("[DEBUG] Rate limit %d reached, sleeping for %s (until %s) before retrying",
+			rlErr.Rate.Limit, retryAfter, time.Now().Add(retryAfter))
+		time.Sleep(retryAfter)
+		rlt.unlock(req)
+		return rlt.RoundTrip(req)
+	}
+
+	rlt.unlock(req)
+
+	return resp, nil
+}
+
+// drainBody reads all of b to memory and then returns two equivalent
+// ReadClosers yielding the same bytes.
+func (rlt *rateLimitTransport) drainBody(b io.ReadCloser) (r1, r2 io.ReadCloser, err error) {
+	if b == http.NoBody {
+		// No copying needed. Preserve the magic sentinel meaning of NoBody.
+		return http.NoBody, http.NoBody, nil
+	}
+	var buf bytes.Buffer
+	if _, err = buf.ReadFrom(b); err != nil {
+		return nil, b, err
+	}
+	if err = b.Close(); err != nil {
+		return nil, b, err
+	}
+	return ioutil.NopCloser(&buf), ioutil.NopCloser(bytes.NewReader(buf.Bytes())), nil
+}
+
+func (rlt *rateLimitTransport) lock(req *http.Request) {
+	ctx := req.Context()
+	log.Printf("[TRACE] Aquiring lock for GitHub API request (%q)", ctx.Value(ctxId))
+	rlt.m.Lock()
+}
+
+func (rlt *rateLimitTransport) unlock(req *http.Request) {
+	ctx := req.Context()
+	log.Printf("[TRACE] Releasing lock for GitHub API request (%q)", ctx.Value(ctxId))
+	rlt.m.Unlock()
+}
+
+func (rlt *rateLimitTransport) isWriteMethod(method string) bool {
```

Comment:
this method I feel more strongly to having just as a little helper and not a method